packages/react-graphql/src/hooks/useQuery.ts

```diff
@@ -0,0 +1,165 @@
+import {
+  OperationVariables,
+  ApolloError,
+  WatchQueryOptions,
+} from 'apollo-client';
+import {DocumentNode} from 'graphql-typed';
+import {useEffect, useMemo, useState} from 'react';
+
+import {AsyncQueryComponentType} from '..';
+
+import {QueryHookOptions, QueryHookResult} from './types';
+import {useApolloClient} from './ApolloContext';
+import useGraphQLDocument from './useGraphQLDocument';
+import {
+  getCachedObservableQuery,
+  invalidateCachedObservableQuery,
+} from './queryCache';
+import {compact, objectToKey, actHack} from './utilities';
+
+export default function useQuery<
+  Data = any,
+  Variables = OperationVariables,
+  DeepPartial = {}
+>(
+  queryOrComponent:
+    | DocumentNode<Data, Variables>
+    | AsyncQueryComponentType<Data, Variables, DeepPartial>,
+  options: QueryHookOptions<Data, Variables> = {},
+): QueryHookResult<Data, Variables> {
+  const {
+    skip = false,
+    pollInterval,
+    notifyOnNetworkStatusChange = false,
+    client: overrideClient,
+    context,
+    variables,
+    fetchPolicy,
+    errorPolicy,
+  } = options;
+
+  const [responseId, setResponseId] = useState(0);
+  const client = useApolloClient(overrideClient);
+  const query = useGraphQLDocument(queryOrComponent);
+
+  const watchQueryOptions: WatchQueryOptions<Variables> = useMemo(
+    () =>
+      compact({
+        context,
+        errorPolicy,
+        fetchPolicy,
+        notifyOnNetworkStatusChange,
+        pollInterval,
+        query,
+        variables,
+      }),
+    [
+      query,
+      pollInterval,
+      notifyOnNetworkStatusChange,
+      context && objectToKey(context),
+      variables && objectToKey(variables),
+      fetchPolicy,
+      errorPolicy,
+    ],
+  );
+
+  const observableQuery = useMemo(
+    () => {
+      if (!query) {
+        return;
+      }
+
+      // eslint-disable-next-line consistent-return
+      return getCachedObservableQuery<Data, Variables>(
+        client,
+        watchQueryOptions,
+      );
+    },
+    [query, client, watchQueryOptions],
+  );
+
+  useEffect(
+    () => {
+      if (skip || !observableQuery) {
+        return;
+      }
+
+      const invalidateCurrentResult = () => {
+        // A hack to get rid React warnings during tests. The default
```

Comment:
as @lemonmade pointed out earlier, he want me to get rid of these.

thou testing is still quite difficult because of this type of hack we needed to do.
https://github.com/facebook/react/issues/14769